lua/starfall/libs_sh/entities.lua

```diff
@@ -1010,14 +1036,68 @@ if SERVER then
 		checkluatype(val, TYPE_NUMBER)
 		ent:SetMaxHealth(val)
 	end
-		
+
 	--- Stops the entity from being saved on duplication or map save.
 	-- @server
 	function ents_methods:doNotDuplicate()
 		local ent = getent(self)
 		checkpermission(instance, ent, "entities.doNotDuplicate")
 		ent.DoNotDuplicate = true
 	end
+
+	--- Fires a bullet.
+	-- @server
+	-- @param table bulletInfo The bullet data to be used. See https://wiki.facepunch.com/gmod/Structures/Bullet
+	-- @param function? callback Function to be called with attacker, traceResult after the bullet was fired but before the damage is applied (the callback is called even if no damage is applied).
+	function ents_methods:fireBullets(bulletInfo, cb)
+		local ent = getent(self)
```

Comment:
I'm not sure what the point of firing bullets from specific entities is. I'd probably make the entity always be instance.entity and rename this function to `game.fireBullets`. Then the SetInflictor in the callback can be removed, and you can make the callback only be added if the user specifies one.

Reply:
if using the chip as the entity firing it will mess with tracer originating from the chip

Reply:
The tracer doesn't use the src and dir?

Reply:
https://wiki.facepunch.com/gmod/Entity:FireBullets see note 2, I also tested to be sure and the tracer was coming out of the chip.

Reply:
What if we disable tracers for this function? Then users will be forced to use the effect library for their own

Reply:
I guess creating their own would be hard since bullet randomness wouldn't be possible to know. Damn

Reply:
I doubt the existing tracers are accurate though since they don't even use Src

Reply:
Is the only reason for making the chip being the firing ent is to avoid having to always set a callback with Setinflictor and perf ? I believe using the tracer provided will always be cheaper than having to make their own using SF, and if the user provide a callback it will be pretty much same perf cost anyway.

Reply:
Not only that, but I'd like this to be in the game library alongside game.blastDamage. This could even be renamed to game.bulletDamage.